New Procedural Macros Crate v0.1

jhwgh1968 opened the pull request:
Update: this is ready for review. It contains:

1. A function decorator to make exposing Rhai functions easier.
2. A module decorator, allowing a Rust module to turn into a Rhai module easily.

Examples are in the top-level crate docs.

### Original Description
An early "sneak peak" of the macro library I'm working on. As the title of the PR suggests, this is not ready for merge. Once you have given your feedback and I have done a little more development, I will have update this with more robust code. Then it will be ready.

This PR creates a new macro library, with all the infrastructure needed to write and maintain the planned procedural macros:

1. A function decorator to make exposing Rhai functions easier.
2. An impl decorator, to expose all public methods easily. (Note: raises minimum version of Rust required by Rhai!)
2. A module decorator, allowing a Rust module to turn into a Rhai module easily.
3. A plugin decorator, which automatically adds things like modules and functions to an engine
   when used.

Only (1) is partly implemented. The syntax is still being developed, and there are a lot of types and cases it doesn't handle yet.

You can look at `macros/tests/test_function.rs` for what a user would have to do _currently_ in order to use the annotation. It's not as nice as I would like, and I will improve it.

Eventually, the vision is something like this for a module:

```rust
#[rhai::module("Advanced Math"]
pub mod AdvancedMath {
        // this type will be "Advanced Math::Point"
        pub struct Point { x: FLOAT, y: FLOAT }

        #[rhai::impl_export(pub)]
        impl Point {
            pub fn new(x: FLOAT, y: FLOAT) -> Point { Point { x, y } }
        }

        #[rhai::impl_export(trait)]
        impl From<(FLOAT, FLOAT)> for Point {
            fn from(pt: (FLOAT, FLOAT)) -> Point { Point { pt.0, pt.1 }
        }
    }

    // this function will be "Advanced Math::euclidian_distance"
    #[rhai::function]
    pub fn euclidian_distance(x: (FLOAT, FLOAT), y: (FLOAT, FLOAT) -> (FLOAT, FLOAT) {
        /* code here ... */
    }
}

fn main() { 
    let mut e = rhai::Engine::new();
    rhai::add_module!(e, AdvancedMath);
}
```

A _plugin_, by the way, is just a module, but with extra flexibility and a couple more behavior hooks (e.g. executing code when it is added to the engine).

I opened this PR, because I am looking for some early feedback. Specifically:

1. The approach: are the items I have planned the way to go?
2. Naming things: the crate name, the attribute names, etc.
3. Organization: since proc macros *must* be their own crate, do you prefer another approach (e.g. workspaces) over my current one (an "embedded" separate crate as an unversioned dependency)?
5. Readability: proc macro rules tie my hands a lot, but is there something you don't understand or think might be difficult to maintain?
6. Direction: even though this is not ready for users of Rhai, I think very little needs to be added to this PR to start implemented standard library functions via annotation. Do you think that would be worth it? What would help that much more narrow use case?

jhwgh1968 commented:
Oh, and when I say:
> A plugin, by the way, is just a module, but with extra flexibility and a couple more behavior hooks (e.g. executing code when it is added to the engine).

I mean that in terms of the syntax that the _user_ sees when using the `#[plugin`] annotation. Internally, it uses the interface I created previously, to allow maximum flexibility in behavior.

schungx commented:
The proc macros are beyond me, but I can comment on your general direction:

>   A function decorator to make exposing Rhai functions easier.

Not sure what you mean.  You mean a more robust form of `Engine::call_fn`?

>    An impl decorator, to expose all public methods easily. (Note: raises minimum version of Rust required by Rhai!)

I suppose this is done on a type level?  Now this would be _cool_.

>    A module decorator, allowing a Rust module to turn into a Rhai module easily.

This is now officially _super_ cool.

>    A plugin decorator, which automatically adds things like modules and functions to an engine when used.

Icing on the cake!

jhwgh1968 commented:
> Not sure what you mean. You mean a more robust form of `Engine::call_fn`?

I meant just the `rhai::function` decorator that adds the function to a Rhai module. It's in the example above.

---

I've made more progress, but now I'm stuck. I have posted a WIP, in the hopes you can tell me what the problem is.

If you:
1. Check out my branch
2. `cd macros && cargo test`

It will run several tests. One of them fails, because the function I am decorating cannot be found.

I presume this is because registering it with `set_fn` means the hash is not presuming a "method receiver" typing. But when I look at the code, I can't tell why method receivers work at all. They seem to be special cased as `set_fn_1_mut` and friends, but I don't see any of that when you do a `register_fn` with a `&mut` parameter closure.

Can you give me any pointers?

schungx commented:
`add_in_place` is not the right name.  It has been registered as `saturating-add` which is not a valid function name.

Changing it to `saturating_add` and then the script to `math::saturating_add(x, 1.0);` will work.  Notice that `1` should be `1.0` to match to a `FLOAT`.

Then finally the answer will be wrong.  All arguments are passed as value, so `x` is a copy.

This is a problem in my code.  I have converted normal function calls to take the first parameter as mutable, but not module-qualified functions.  I'll fix that.

schungx commented:
I've fixed it.  Now passing a variable as first argument to a function with `&mut` will mutate that first argument.

jhwgh1968 commented:
Thanks! I got that case working, and did a push just to update things.

---

I am now thinking I'll implement 1 and 3 (functions and modules) since they will be working very closely together. Then I will rename the PR once I am ready for merge.

Since making that decision, I am looking at your initial comments again. Unless I missed something, they seems to mostly answer my first point:

> The approach: are the items I have planned the way to go?

It seems like your answer is an enthusiastic yes. That's good!

Do you have any other comments in regards to the other points? Particularly naming (the sooner I can get that fixed the better), and internal uses (whether module + function decorators will let you re-write some of the Rhai "core library" to be more readable).

I fully expect these attributes will co-exist with `#[cfg(feature = ...)]`, but I haven't tested exactly how the parsing works. You may need an extra pair of braces or two, in the worst case.

schungx commented:
I haven't really checked out the "usage" experience yet...  I'd suggest you get a stable version working, then we'll spend some time using the new interface and iterating it to see if anything else needs update.

schungx commented:
Not very sure what `imp_export(pub)` and `imp_export(trait)` is necessary for, since you can always treat `Point` as a simple Rust type.  What extra code is added with these macros?

jhwgh1968 commented:
> Not very sure what `imp_export(pub)` and `imp_export(trait)` is necessary for, since you can always treat `Point` as a simple Rust type. What extra code is added with these macros?

The actual methods that take a receiver.

Without those attributes on the impl blocks, Rhai will understand a Point, but only its fields. As written above, since `x` and `y` are not marked `pub`, it would be a completely opaque type. You would have to write separate Rust functions to use or create it:

```
#[rhai::function]
pub fn new_point(x: FLOAT, y: FLOAT) -> Point { ... }
#[rhai::function]
pub fn cast_point(tuple: (FLOAT, FLOAT)) -> Point { ... }
```

With the attributes on the impl blocks, the methods in the contents of those blocks will be re-exported to Rhai as well, meaning that `Point::new` and `Point::from` above will be exported.

I am doing it this way, as the best compromise between three alternatives:
1. Decorate each method to export with `rhai::function`. This struck me as a lot of typing, especially during development. I may add an option for this later.
2. Always export `pub fn`s on every type that was exported to Rhai. This is the least typing, but may not be what users want. I could see having some "Rust only" functions that are still pub for scoping reasons.
3. Select the export behavior on each impl block for its contents. This is what I did. While "export everything marked pub" is the example, I am also thinking about stuff like "export everything with prefix". 

Hopefully that answers your question.

jhwgh1968 commented:
Oh, and in case you are wondering, `export(pub)` means "export everything I've marked pub." By contrast, `export(trait)` means: export all the methods on this trait, _even if they have a default implementation not defined in this block_. That is why it's a separate tag.

schungx commented:
Now I understand.  Good design!

Question: Do I have to write property accessors for each property or will they be auto-generated?

Like `point.x` and `point.y`...

jhwgh1968 commented:
> Question: Do I have to write property accessors for each property or will they be auto-generated? Like `point.x` and `point.y`...

I was thinking that all `pub` fields should get accessors by default, with an opt-in way to turn them off. This is because `pub` fields in Rust code are pretty rare. Basically, it means that the type has no invariants based on the field.

When I have seen this, either it is an "output" type which is purely structured for convenience (i.e. it loses all invariants after being produced, expecting the caller will destructure it) or it is `Copy` type that is nothing more than the sum of its parts.

My Point example could illustrate either one, depending on its uses. It depends on what the user wants to do with e.g. NaNs and Infinities.

All that said, this aspect is probably one of the last things I'll implement. I personally think the case is somewhat marginal, and I have not looked at all into the Rhai side of hooking into the dot syntax.

schungx commented:
I think if there is a `pub` Rust field, make both get and set property accessors.

If there is a function of the correct signature as a field, automatically register it as accessor.

```rust
struct Point { x: i64, y: i64 };

impl Point {
    // this should be registered as getter 'x'
    pub fn get_x(&self) -> i64 {
        self.x
    }
    // this also
    pub fn x(&self) -> i64 {
        self.x
    }
    // this should be registered as setter of 'x'
    pub fn set_x(&self, value: i64) {
        self.x = value;
    }
}

// if the type implements Index, register an indexer
impl Index<i64> for Point {
    ...
}
```

jhwgh1968 commented:
> I think if there is a `pub` Rust field, make both get and set property accessors.

Agreed.

> If there is a function of the correct signature as a field, automatically register it as accessor.

I'm a little nervous about that, however. There would need to be some heuristics involved due to getters doing casts, such as:

```
struct Point { x: f32, y: f32 }
impl Point {
    pub fn get_x(&self) -> noisy_float::R32 {
        R32::new(self.x).unwrap()
    }
}
````

jhwgh1968 commented:
More progress has been made, but another sticking point. Namely, constant exports aren't working.

Here is the failing test:
```
pub mod one_fn_and_const_module {
    use rhai_macros::module as rhai_module;

    #[rhai_module]
    pub mod advanced_math {
        use rhai::FLOAT;

        pub const MYSTIC_NUMBER: FLOAT = 42.0 as FLOAT;

        pub fn euclidian_distance(x1: FLOAT, y1: FLOAT, x2: FLOAT, y2: FLOAT) -> FLOAT {
            ((y2 - y1).abs().powf(2.0) + (x2 -x1).abs().powf(2.0)).sqrt()
        }
    }
}

#[test]
fn one_fn_and_const_module_test() -> Result<(), Box<EvalAltResult>> {
    let mut engine = Engine::new();
    let m = crate::one_fn_module::advanced_math::rhai_module__generate();
    let mut r = StaticModuleResolver::new();
    r.insert("Math::Advanced".to_string(), m);
    engine.set_module_resolver(Some(r));

    assert_eq!(engine.eval::<FLOAT>(
        r#"import "Math::Advanced" as math;
           let m = math::MYSTIC_NUMBER;
           let x = math::euclidean_distance(0.0, 1.0, 0.0, m);
           x"#)?, 41.0);
    Ok(())
}
```

My procedural macro attribute adds a `generate()` function to that module, which expands to:

```
pub fn rhai_module__generate() -> Module {
    let mut m = Module::new();
    m.set_fn("euclidian_distance", FnAccess::Public,
             &[core::any::TypeId::of::<FLOAT>(),
               core::any::TypeId::of::<FLOAT>(),
               core::any::TypeId::of::<FLOAT>(),
               core::any::TypeId::of::<FLOAT>()],
             rhai::plugin::CallableFunction::from_plugin(euclidian_distance__Token()));
    m.set_var("MYSTIC_NUMBER",
              rhai::Dynamic::from(42.0 as FLOAT));
    m
}
```

That code seems okay to me, yet it cannot find the variable in the module when the test runs. Do you know why?

schungx commented:
Strange.  The code looks correct...

jhwgh1968 commented:
Whoops, I forgot to paste the error! It's my "favorite" error:

```
Error: ErrorVariableNotFound("math::MYSTIC_NUMBER", 2:26)
```

jhwgh1968 commented:
> If you take away this statement, will the next line work?

Hmm.... if I manually set `let m = 42.0` instead, it can't find the function either. That's weird.

jhwgh1968 commented:
Okay, I still can't spell "euclidean". :laughing: 

But even after that typo fix, it still doesn't work. Which doesn't surprise me, because I can spell "mystic".

schungx commented:
So it is still not finding `math::MYSTIC_NUMBER` right?

jhwgh1968 commented:
> So it is still not finding `math::MYSTIC_NUMBER` right?

Correct.

schungx commented:
Can I see the expanded code that comes from `MYSTIC_NUMBER`?

schungx commented:
`m.set_var("MYSTIC_NUMBER", 42.0 as FLOAT);`

This is what you need.  No need to convert it to a `Dynamic`.  Otherwise, that `Dynamic` will be wrapped up as a trait object.

Try it.  Not sure if it solves the problem...  Technically speaking, it still should find the variable, just with the wrong value...

schungx commented:
I just tried your original code in the `master` branch inside one of the tests, and it finds the variable fine...

jhwgh1968 commented:
I fixed it. It was another one of my famous copy-past errors.

The only step left before I declare it "ready to merge" is to expose these macros through the Rhai crate itself, rather than through direct exports. This is needed both to keep the interface clean, and to allow other utility macros to clean up things like the `__Token` and `__generate()` calls I have in the tests. (Procedural macro crates can contain _only_ procedural macros. Not even regular macros.)

---

As a side note, I built an optimized binary and dumped the assembly for the macro-generated `PluginFunction::call()` for the `euclidean_distance` function. The results are not quite as good as I had hoped, but I doubt much better performance could be gained by a different approach.

The actual function body of `euclidean_distance` is extremely tight: a mere 10 instructions on Intel processors with SSE (virtually all 64-bit chips that are still running). That block is then is then inlined into the `PluginFunction::call()` template (which I would fully expect), and surrounded by the macro-generated glue of approximately 250 instructions.

My quick reading suggests about 200 of those instructions are hit on the happy path (where the function is called and returns a value). And that's after I specifically asked the compiler to inline calls to `Dynamic::as_ref<T>`.

In other words, calling this function once is not very expensive in the course of a program. But doing it in a loop over 1,000 data points will cause overhead users will probably notice.

I'm assuming that's not a big deal, as speed is a secondary (if not tertiary) concern, and the usability wins are definitely worth it. I also doubt it's that much worse than the current system of "packages" it will replace, though I haven't measured that.

schungx commented:
I would assume that the currently packages will be exactly the same, as they run through the same `set_fn` calling mechanism.  Rhai has pretty high overheads, not the least to convert to and from `Dynamic` values instead of straight types, and these dynamic types defeat the compiler's ability to optimize.  This is probably why a dynamic interpreted language is slower than native...

From your description, we can probably spend some time on optimizing the hot path of the calling mechanism.

If you can paste here a few expanded macros for the function registration, I'll see if there is something we can do better...  Especially, for proc macros, you can actually compile in the actual type instead of a `Dynamic`...  but of course the return value must be a `Dynamic`.

As for exposing the macros through the Rhai crate itself, not sure what you mean.  Can we just put in a bunch of `pub use` statements in `lib.rs`?

jhwgh1968 commented:
> From your description, we can probably spend some time on optimizing the hot path of the calling mechanism.

I hope to play around with Compiler Explorer (as Matt Godbolt requested we call it in a recent presentation) and see what I can do.

My current thought is to create a new method on `Dynamic` specifically for the idiom my generated code will do most often. Something like:

```pub fn as_clone<T: Clone>(&self) -> Option<T>````

Instead of writing a `match` statement based on the value of the `Dynamic`, the match statement would be based on the type ID of `T`, and the content of `self` would be accessed directly where possible.

I hope that will enable dead code elimination to make the happy path shorter for simple cases, and the same if `T` is held in a `Box<T>`.

> As for exposing the macros through the Rhai crate itself, not sure what you mean. Can we just put in a bunch of `pub use` statements in `lib.rs`?

That's exactly what I meant. I just haven't done that yet :smile: 

There are also some other macros I'd like to write for ease of use, too.

jhwgh1968 commented:
I just tried it quick, and discovered it is _really_ difficult to create a single that compiles this code:

```
fn main() {
    let d = Dynamic::from(42.0 as FLOAT);
    println!("The magic number is {}", d.downcast_ref::<FLOAT>().unwrap());
}
```

If you could create a Compiler Explorer link with that code -- and the hundreds of pasted Rhai lines above it -- that would help a lot.

jhwgh1968 commented:
As I had hoped, CE confirms that this code optimizes to just a value copy (with slightly different version of `Dynamic`):

```
    pub fn downcast_copy<T: Copy + 'static>(&self) -> Option<T> {
        if TypeId::of::<T>() == TypeId::of::<i32>() {
            self.as_int().map(|i| {
                let ptr = &i as *const i32;
                unsafe { *(ptr as *const T) }
            })
        }
        else if TypeId::of::<T>() == TypeId::of::<f32>() {
            self.as_float().map(|i| {
                let ptr = &i as *const f32;
                unsafe { *(ptr as *const T) }
            })
        }
        else if TypeId::of::<T>() == TypeId::of::<char>() {
            self.as_char().map(|i| {
                let ptr = &i as *const char;
                unsafe { *(ptr as *const T) }
            })
        }
        else if TypeId::of::<T>() == TypeId::of::<bool>() {
            self.as_bool().map(|i| {
                let ptr = &i as *const bool;
                unsafe { *(ptr as *const T) }
            })
        }
        else { None }
    }
```

I'll see what the new instruction stream looks like, once I implement it for actual `Dynamic`.

jhwgh1968 commented:
The answer is, it's an improvement, but it's a bit hard for me to follow, because I don't know what Rust's memory management tricks are in raw assembler.

There are way more branches (due to my use of an `Option::or_else` block and a number of nested `unwrap` checks), but branch prediction should handle those easily. If the branch checks were zero cast, the happy path is down to about 80 instructions before the body of the function and 6 instructions after.

Much better!

schungx commented:
So, where do you find the bottleneck being?  Is it creation of a `Dynamic` from a primary value?  Or is it the other way round, copying a `Dynamic` into a primary value?

jhwgh1968 commented:
I'm not sure there is a "bottleneck", exactly... but I will try to answer the question I _think_ you mean.

I believe the least efficient chunk of code -- which also has the highest risk of CPU stalls -- is `Dynamic::downcast_ref<T>`. It's one giant `match` statement on the Dynamic passed as `self`, and that is _inherently_ hard to optimize.

The value of the `Dynamic` passed into any call really could be any variant. In even a medium-sized script, there is no pattern for the branch predictor to find (e.g. "predict it's a float variant" or "predict it's a Box variant"). The likelihood of mispredictions and stalls is high. Thus, I wrote `Dynamic::downcast_copy`, which tries to create a much shorter happy path for simple types.

Once Rust's type specialization and LLVM's dead code elimination are done with it, the code generated for each float argument pulled from the array in the preamble for `euclidean_distance` is equivalent to:
```
let argN: FLOAT = args[N].as_float().or_else(|| *Dynamic::downcast_ref<FLOAT>(self)).unwrap();
```

(I put the `downcast_ref` in there, because I presumed this code should unpack a `Box<f32>` someone was using for some reason, rather than considering it a type mismatch. It won't affect the happy path, regardless.)

As best I can tell, the optimization in my simpler example is preserved. Despite how weird the source code for `downcast_copy` itself looks, LLVM realizes I am just copying a value from one place to another, and generates assembly to do that.

Mutable first arguments still require a downcast:

```
let arg0: &mut MyType = args[0].downcast_mut::<MyType>(self)).unwrap();
```

But since they can only be first arguments, and that is the only time my code calls it, the branch predictor has a better chance to correctly predict the `match` statement in `downcast_mut`. I think.

I have not traced through the branches all the way down in my current version, But each argument seems to generate about 16 instructions in total. That includes four conditional branches which won't get taken by the happy path (e.g. `unwrap()` checks), and memory and register scheduling. The rest of the overhead seemed fixed (stack handling and setting up for the function body itself).

Unless you have another clever idea, I think that is pretty close to optimal.

schungx commented:
I think I understand what you mean.  You're right.  `downcast_ref` and `downcast_mut` are in a _lot_ of hot paths within the code, not just for function interop.

Converting it from a giant `match` statement to a chain of `if` statements will allow inlining when the cast is to a specific type.  This may actually have performance benefits as well.

schungx commented:
OK, I followed your idea and changed `cast`, `try_cast`, `from`, `downcast_ref` and `downcast_mut` from a giant match statement into individual `if TypeId::of::<T>() == TypeId::of::<INT>()` blocks.

It turns out to have significant regression on all benchmarks, some by more than 50%.  In other words, all benchmarks are _slower_ after making this change.

Not sure what the problem is...  maybe Rust has special logic to handle match statements on enums?

schungx commented:
I used GodBolt and tried both versions of `try_cast`, one with a giant `match` statement and one with `if` statement blocks.

The code generated do not appear to have any significant differences.

The giant `match` statement optimizes down to a single compare instruction on the `enum` tag, just checking for the correct type variant.  All other variants compile down to `None`, essentially non-branch becomes `None`.

This is essentially the same as pruning all `if` statement blocks of all other types, except for the correct one and then matching on the single variant alone.

Did you use `-C opt-level=3` to turn on optimizations?  None of my casts turn out to have more than a few instructions each...

jhwgh1968 commented:
> Converting it from a giant `match` statement to a chain of `if` statements will allow inlining when the cast is to a specific type. This may actually have performance benefits on other parts of the code base as well.

To be clear, I am not sure if that's true _in general_. I agree, the compiler is very smart about folding and eliminating branches. I was looking _at my use case in a particular_.

When I marked `downcast_ref` as `#[inline]` but made no other changes, the amount of code inlined into `euclidean_distance` above the body of the function was significantly larger than the current PR (around 200 instructions vs about 80). I didn't investigate why, but my guess is that the control flow graph could not (or would not) be optimized the same.

This is not because the compiler missed something, but because `dynamic_copy` already knows that all the types are `Copy`. When looking at the enum in its entirety, some cases exist where boxes are moved out of. That means destructors have to be scheduled before function return.

It could also be that the happy path is similar, and I just read it too fast.

schungx commented:
> I think this is not because the compiler missed something, but because `dynamic_copy` already knows that all the types are `Copy`. When looking at the enum in its entirety, some cases exist where boxes are moved out of. That means destructors have to be scheduled, for example.

This could be it.  If all types are `Copy` then I suppose more optimizations can be made.

> When I marked downcast_ref as #[inline] but made no other changes

The compiler probably didn't want to inline the function because it is very large.  However, since most branches will be compiled away as dead code, actually the function becomes a simple statement.

This is probably the case where `#[inline]` helps a great deal.  I'll put `#[inline]` on all these functions and rerun the benchmarks.

schungx commented:
BTW, for your `PlugInFunction::call` method, I suggest dropping the `Position` argument.  It is only used for error reporting, and there is no point to always pass a small struct which is only used in the error path.

Use `Position::none()` as the position for your errors.  `Engine` will set the position to the correct value after the call if there is an error.

jhwgh1968 commented:
> Use `Position::none()` as the position for your errors. `Engine` will set the position to the correct value after the call if there is an error.

I remember you said that before, but why didn't I do it then...

Ah yes, that's why. It's export is `pub(crate)` meaning that my separate proc macros crate can't see it.

Since I've marked several things that were `pub(crate)` as `pub`, I'll add that to the list.

schungx commented:
No prob!

jhwgh1968 commented:
I'm currently working my way to string handling for decorated function, i.e. autocasting to and from `rhai::ImmutableString` when the caller takes a `&str`.

In so doing, I have worked through several edge cases, and I would like to double-check them with you:

1. The return type cannot be a reference or a pointer. Every return type must be owned. In principle, `& 'static` references would be okay, but this is mostly to allow applying `Dynamic::from` blindly and easily.
1. The first parameter may be a mutable reference to any type T.
1. Strings passed in from Rhai are accepted as `&str`, and strings returned to Rhai are returned as `String`. This will be handled by custom logic converting to and from Rhai's native `ImmutableString`.
1. Array and Map types are passed in from Rhai as a mutable reference directly (e.g. `rhai::Array`), and can appear in any argument position. They can also be returned as owned types, per rule 2.
1. Passing in owned or mutably borrowed (but not shared) `Dynamic` is allowed in any position. 1. All types passed in from Rhai that do not follow the above must be `Copy` types.

I expect that last rule might seem quite strict at first. However, I cannot think of any cases not covered by the others above, and it would vastly simplify the logic in the macro. (And _maybe_ generate more optimal code, per our discussion above.)

What do you think?

schungx commented:
1. That's right.  Currently all functions registered with Rhai must be `'static`.  That is to avoid putting a lifetime on the engine.

2. Correct.

3. There is code in `fn_register.rs` that detects if a function accepts `&str`.  If it does, it casts the `ImmutableString` value into `&str` before passing to the function.  The function can return `String` or `ImmutableString`.  Either way, the end result is `ImmutableString`.

4. No.  Arrays and Maps are cloned, unless they are in the first position which is `&mut`.  There are no references in Rhai.  Deep inside, the raw function takes `&mut Dynamic` references to all the copied values.  In other words:

    First `&mut` -> reference is the same
    First non-mut -> reference to cloned copy
    Other arguments -> reference to cloned copy

5. Similar to No.4.  There is no way to actually pass `&Dynamic`.  If the first position is `&mut` it is a mutable reference to the value.  Otherwise, it is a mutable reference to a _copy_ of the value.

6. All types must be `Clone`.  They do not need to be `Copy`.

jhwgh1968 commented:
I will adjust to that.

I was planning to do the casting myself in item 3, since I wasn't sure whether `register_fn` is used eventually when I do a `set_fn` in a module. If I don't have to, so much the better! Testing will tell me that pretty quickly.

I also replaced my `downcast_copy` with `downcast_clone`. It generates more instructions for my `euclidian_distance` example, but I'm now sure it's not any worse. I found something I missed before: an "ultra fast" path through the code, which is optimized for types that are `Copy`. (At least, when you build with `codegen-units=1`.)

In case you are wondering, yes I would like this to be a utility method on `Dynamic`, even thought it only consists of:

```
#[inline(always)]
pub fn downcast_clone<T: Variant + Clone>(&self) -> Option<T> {
    self.downcast_ref::<T>().map(|t| t.clone())
}
```

Wrapping it in a utility function makes it much easier for me to have a useful error message _pass through the macro_ when the user doesn't know about the restrictions you laid out:

```
error[E0277]: the trait bound `NonClonable: std::clone::Clone` is not satisfied
  --> src/user_code.rs:198:26
   |
98 |     pub fn my_function(input: NonClonable) -> bool {
   |                               ^^^^^^^^^^^ the trait `std::clone::Clone` is not implemented for `NonClonable`

```

I consider that kind of usability key to my implementation.

schungx commented:
> I was planning to do the casting myself in item 3, since I wasn't sure whether `register_fn` is used eventually when I do a `set_fn` in a module. If I don't have to, so much the better! Testing will tell me that pretty quickly.

I think it is the other way round - `register_fn` uses `set_fn`.  So I think you may have to copy the mechanism in `register_fn`...

jhwgh1968 commented:
Alright, here we go again...

Why does my string example in `tests/test_functions.rs` throw the error:

```
Error: ErrorFunctionNotFound("io::write_out_str", 2:24)
```

The code:

```
mod raw_fn_str {
    use rhai_macros::function as rhai_function;

    #[rhai_function]
    pub fn write_out_str(message: &str) -> bool {
        eprintln!("{}", message);
        true
    }
}

#[test]
fn raw_fn_str_test() -> Result<(), Box<EvalAltResult>> {
    let mut engine = Engine::new();
    engine.register_fn("get_mystic_number", || { 42 as FLOAT });
    let mut m = Module::new();
    m.set_fn("write_out_str".to_string(), FnAccess::Public,
             &[std::any::TypeId::of::<&str>()],
             CallableFunction::from_plugin(raw_fn_str::rhai_fn__write_out_str::Token()));
    let mut r = StaticModuleResolver::new();
    r.insert("Host::IO".to_string(), m);
    engine.set_module_resolver(Some(r));

    assert_eq!(engine.eval::<bool>(
        r#"import "Host::IO" as io;
           let x = io::write_out_str("hello world!");
           x"#)?, true);
    Ok(())
}
```

Is it supposed to pass `ImmutableString` to the `set_fn` call?

schungx commented:
>     m.set_fn("write_out_str".to_string(), FnAccess::Public,
>              &[std::any::TypeId::of::<&str>()],
>              CallableFunction::from_plugin(raw_fn_str::rhai_fn__write_out_str::Token()));

When passed to `set_fn` it must be `ImmutableString`, not `&str`.  You do your conversion outside of the module...

jhwgh1968 commented:
> When passed to `set_fn` it must be `ImmutableString`, not `&str`

That did the trick. Thanks!

The code it actually generates is:
```
let arg0 = args[0usize]
    .downcast_clone::<rhai::ImmutableString>()
    .unwrap();
Ok(rhai::Dynamic::from(write_out_str(&arg0)))
```

I tried `borrow` and `as_str`, but those required trait imports. Thank goodness you implemented `Deref` as well!

jhwgh1968 commented:
Now I'm having more trouble with `Dynamic`. And it's rather ominous trouble.

Here is my current test case -- a similar pattern to my future plans for the Rhai language:

```
mod mut_str_collide {
    use rhai_macros::function as rhai_function;

    #[rhai_function]
    pub fn get_stdout() -> std::io::Stdout {
        std::io::stdout()
    }

    #[rhai_function]
    pub fn write_out_str(stream: &mut std::io::Stdout, message: &str) -> bool {
        write!(stream.lock(), "{}", message).is_ok()
    }
}
```

I fully expect this to work. I'm taking the stdout handle, giving ownership to Rhai, and then only ever accessing it again through a mutable borrow. Given that my current macro rules _only_ allow returning _any_ type by value, many custom types created in Rust and exposed to Rhai will be used in this way.

However, it doesn't work:

```
error[E0277]: the trait bound `std::io::Stdout: std::clone::Clone` is not satisfied
   --> tests/test_functions.rs:99:5
    |
99  |     #[rhai_function]
    |     ^^^^^^^^^^^^^^^^ the trait `std::clone::Clone` is not implemented for `std::io::Stdout`
    | 
   ::: /home/cryptkeeper/Desktop/Software/ytdl-replacement/rhai/src/any.rs:346:30
    |
346 |     pub fn from<T: Variant + Clone>(value: T) -> Self {
    |                              ----- required by this bound in `rhai::Dynamic::from`
```

I was surprised, because the only place this could be coming from is the `Dynamic::from` around the return value. I thought `Variant` was supposed to cover everything, so you should be able to do `Dynamic::from` on any type.

Apparently not. It has to be `T: Variant + Clone`.

That is a very serious limitation, and is quite crippling to my future use case. 

I tried to fix this myself, but two or three impl's I've tried conflict with others you already have. Do you have a way out of this?

schungx commented:
Not easily.  The reason why all `Dynamic` data types must implement `Clone` is because all arguments are passed by value (i.e. copies).

If we don't want this, then all arguments must be passed by reference, and because of historical reasons, since the first argument must be either a clone or `&mut`, all argument references must be `&mut`.   However, if the arguments reside in the `Scope`, you can ever only get one mutable reference to one variable, not multiple...

Any solution to resolve this will be major surgery.  That's why the `ImmutableString` type is introduced.  Otherwise, we'd just be using `&str` everywhere.

Now if you really want to pass a non-`Clone` type as argument, you have to make a _wrapper_ type that wraps it in an `Rc` (or `Arc` depending on `sync`).  `Rc<T>` or `Arc<T>` is automatically `Clone`.

If that type needs to be mutable, then you can further wrap the type in a `RefCell` or `Mutex` (depending on `sync`).

Now, come to think of it, many of the functions you plan to write affect the external environment -- e.g. writing to a stream.  Rhai is sandboxed, meaning that it cannot affect the running environment if it is not declared mutable.  Therefore, you'd be having trouble with non-pure functions one way or another.

One way to deal with this is to auto-generate a wrapper type with your procedural macros if the type is not `Clone`?

schungx commented:
For example, in your case it would be something like:

```rust
pub type StdoutWrapper = Shared<RefCell<std::io::Stdout>>;

pub fn write_out_str(stream: StdoutWrapper, message: &str) -> bool {
    write!(stream.borrow_mut().lock(), "{}", message).is_ok()
}
```

jhwgh1968 commented:
> The reason why all `Dynamic` data types must implement `Clone` is because all arguments are passed by value (i.e. copies). [...] However, if the arguments reside in the `Scope`, you can ever only get one mutable reference to one variable, not multiple...

I simply don't understand how those rules (which I do not wish to change, to be clear) forbid what I have in mind.

Looking at the implementation of `Dynamic`, it can contain:

```
pub enum Union {
    Unit(()),
    Bool(bool),
    Str(ImmutableString),
    // ...
    FnPtr(Box<FnPtr>),
    Variant(Box<Box<dyn Variant>>),
}
```

If I'm reading this correctly, `Variant` is based on `Any` with a lifetime restriction, which means it can be almost any type. In particular, it does _not_ have to implement `Clone`.

This suggests to me that I _should_ be able to implement:

```
impl<T: Variant> From<Box<T>> for Dynamic
```

This will basically "box up the box", and give the only reference to my `T` to the `Dynamic` I create.

If I called `Dynamic::downcast_clone` on it, or many of the other accessors, they would return `None` or fail, resulting in a runtime error. But what if code never did _anything except_ call `downcast_mut` on it? That is, it was always passed as `&mut` in the first argument to everything that called it?

Creating an `ImmutableString` type make sense, because that calling pattern is overly restrictive for the concept what a string is -- in either Rust or Rhai. But there are other cases -- such as host  resource access via a [proxy object](https://en.wikipedia.org/wiki/Proxy_pattern) -- where that is all you need to do with it. In those cases, it would be just fine.

What am I missing that prevents this?

schungx commented:
> If I called `Dynamic::downcast_clone` on it, or many of the other accessors, they would return `None` or fail, resulting in a runtime error. But what if code never did _anything except_ call `downcast_mut` on it? That is, it was always passed as `&mut` in the first argument to everything that called it?

Hhhmmm... you raise a good point.  If a type is ever only called in the first `&mut` position, then technically speaking it need not be `Clone`...

It was a historical implementation direction that is going to be quite tricky to change.  Actually, removing the `Clone` requirement on `Dynamic` opens it up to massive benefits, not to mention that we can actually put references into it.

schungx commented:
OK, removing `Clone` hits on the first problem.  The `RegisterFn` trait conflicts with itself because, obviously, `&mut T` is a `T` if `T` is not `Clone`.  The surgery is deeper than it looks

Basically, the plan of attack is:

1. Remove `Clone` bound attached to `Variant`
2. Make `Dynamic` non-clonable
3. When cloning `Dynamic`, use a `try_clone` method instead.
4. Script terminates with a `NonClonableDynamic` error or something, if a non-clonable value is passed as a value parameter.

jhwgh1968 commented:
> The `RegisterFn` trait conflicts with itself because, obviously, `&mut T` is a `T` if `T` is not `Clone`.

That general kind of problem is why I tried to implement a conversion to `Dynamic` from `Box<T>` rather than any `T` itself. The error message I got was about some type in fn_native that I didn't dig into. I don't know if it's the same root cause or not.

> I can't see how this can be done easily within the next few days. I suggest you also think of a "Plan B" which involves auto-generating wrapper objects for interop with Rhai.

The biggest obstacle to any Plan B is procedural macro limitations. I can't make decisions at macro generate time based on the context of `T`, e.g. does it implement `Clone` or not. This is why my approach is "brute force": generate the code that is _necessary_ to match Rhai's calling convention, and if something doesn't fit, let the Rust compiler throw (cleverly spanned) type errors at the user.

Thus, the only Plan B approach I can think of is a new Dynamic conversion. Something like: add a new called `Dynamic::from_nonclonable` which would put the type into an internal `Arc` or `Rc` depending on `feature = "sync"`. It would support clone by implementing a new enum value of `Union` which would call `Arc::clone` on the inner value.

I will try this approach in the next several days.

jhwgh1968 commented:
> > The `RegisterFn` trait conflicts with itself because, obviously, `&mut T` is a `T` if `T` is not `Clone`.
> 
> That general kind of problem is why I tried to implement a conversion to `Dynamic` from `Box<T>` rather than any `T` itself. The error message I got was about some type in fn_native that I didn't dig into. I don't know if it's the same root cause or not.
> 
> > I can't see how this can be done easily within the next few days. I suggest you also think of a "Plan B" which involves auto-generating wrapper objects for interop with Rhai.
> 
> The biggest obstacle to any Plan B is procedural macro limitations. I can't make decisions at macro run time (i.e. code generate time) based on the context of `T`, e.g. does it implement `Clone` or not. This is why my approach is "brute force": generate the code that is _necessary_ to match Rhai's calling convention, and if something doesn't fit, let the Rust compiler throw (cleverly spanned) type errors at the user.
> 
> Thus, the only Plan B approach I can think of is a new Dynamic conversion. Something like: add a new method called `Dynamic::from_nonclonable` which would put the type into an internal `Arc` or `Rc` depending on `feature = "sync"`. It would support clone by implementing a new enum value of `Union` which would call `Arc::clone` on the inner value.
> 
> I will try this approach in the next several days.

jhwgh1968 commented:
I've not gotten it working yet, but I'm close, at least.

Please review [this commit](https://github.com/schungx/rhai/pull/14/commits/6f44f7c82811478c4bc3cf80bb3316da65c1750d) and see if you like this approach.

schungx commented:
Yeah, the original design of the `RegisterFn` trait imposes a significant restriction.  For example, we cannot have `&T` parameters, only `T` and `&mut T` because `&T` will conflict with `T`.  Without `Clone`, `&mut T` also conflicts with `&T`.

This cannot be resolved until Rust has specifications, which is taking forever to land in the language.

I like your idea of `Dynamic::from_non_clonable` though... I'll review your code.

schungx commented:
This PR https://github.com/jonathandturner/rhai/pull/203 does very similar things by introducing a shared `Dynamic` variant, for an entirely different usage.

jhwgh1968 commented:
I looked it over, and I think the author of jonathandturner/rhai#203 is mostly correct in their analysis. However, their proposal is a bit overkill for all I am trying to do, which is satisfied by simply `Arc<dyn Any>`. I am also somewhat troubled by their current PR, which radically alters the `downcast` family of functions I rely on. But that may not be in their next version.

Since this is an area of active investigation by others, I will defer my "proxy object" use case until later, and narrow my focus to try and get this PR merged.

To ask a question I asked before: what do you need to re-write the current "Rhai std library" as Rust modules with my decorators? As far as I can tell:

* A `#[rhai_module]` decorator applied to a rust module which does the macro generation
* Proper exposure of Rhai compatible types according to the closures you can write for `register_fn`
* The ability to expose constants of Rhai compatible types
* An easy way to add modules generated by the macro to an engine instance

The last one is the only one I haven't done yet. If I get that done, would you have time to review this and then try to re-write parts of "Rhai std" to kick the tires?

schungx commented:
> The last one is the only one I haven't done yet. If I get that done, would you have time to review this and then try to re-write parts of "Rhai std" to kick the tires?

Yes definitely!

I think the first step is to design a namespace tree for standard Rhai functions and behaviors.  Sorta like a tech tree in games.  And what the namespaces will look like.  I'll start giving it some thoughts, but you can also give ideas.  You have a much more ambitious goal of porting much of Rust's std library to Rhai, while I'm satisfied with simple utility functionalities for common types.

schungx commented:
> However, their proposal is a bit overkill for all I am trying to do, which is satisfied by simply `Arc<dyn Any>`

I've suggested to the author that it is best to move the type to `Arc<RwLock<dyn Any>>` instead of `Arc<RwLock<Dynamic>>`.

I think you're going to need the lock, because for many objects you'll want it mutable.  For example, a network stream, file handle or database adapter... if you write to any of these, unless they are language built-in (like `println!`), you're going to need a mutable reference to them, and you won't get a `&mut T` from an `Arc<T>` alone...

jhwgh1968 commented:
> I think you're going to need the lock, because for many objects you'll want it mutable.

I was thinking simply using `Arc::get_mut` would work, as the `Arc<T>` would be copied around, but the old copies would go out of scope quickly. But since I couldn't get it working, I can't actually prove this is true. That is the other reason I am dropping this for now.

> I think the first step is to design a namespace tree for standard Rhai functions and behaviors. Sorta like a tech tree in games. And what the namespaces will look like. I'll start giving it some thoughts, but you can also give ideas. You have a much more ambitious goal of porting much of Rust's std library to Rhai, while I'm satisfied with simple utility functionalities for common types.

Maybe "the standard library" is bigger than I was thinking, or not the correct term. I was just thinking the giant list of `register_fn` calls that is always done when setting up a new `Engine` to create behavior on the basic types (int, float, string, etc).

It seems like you already have an organization for that, with your `src/packages` directory. I imagine it's just a matter of changing some `def_package!` macro calls to `#[rhai_module] pub mod` blocks.

Not only would that be a good test of the code generator itself -- did any edge cases get hit that I forgot about? -- but the savings in boilerplate would be significant. It would also not require some of my more ambitious ideas, like selective exports -- which are turning out to be harder than I originally thought in the current design.

schungx commented:
> I was thinking simply using `Arc::get_mut` would work, as the `Arc<T>` would be copied around, but the old copies would go out of scope quickly. But since I couldn't get it working, I can't actually prove this is true. That is the other reason I am dropping this for now.

Not necessarily.  Within Rhai there may be references hanging around.  That's why the `Arc` is necessary, otherwise it can simply be owned if you can guarantee exclusive access.

`Arc::get_mut` can fail and return `None`.  You need the lock.

schungx commented:
> It seems like you already have an organization for that, with your `src/packages` directory. I imagine it's just a matter of changing some `def_package!` macro calls to `#[rhai_module] pub mod` blocks.

Yes, I think the standard packages is ripe to immediately move to plugins, and in the way test out the system!

However, I don't think we should stop at just implementing these packages.  Your macros can potentially make it a piece of cake to very quickly expand the standard library to cover many other functions and features.

jhwgh1968 commented:
Alright, it's ready for a full review!

If you look at the `cargo doc` for the `codegen` library (yes, it's been renamed) you can see the two examples of how to use it.

This covers functions and modules, but not types or `impl` blocks. Based on looking through your packages, it should be enough for a good start.

schungx commented:
Looks like `plugins.rs` has an error in `no_std`.  Should `use crate::stdlib::any::TypeId` instead of `std::any::TypeId`.

jhwgh1968 commented:
Thanks, @schungx. Fixed.

I am also testing the codegen crate on nightly only for now. My `trybuild` tests give different (though logical) results on stable. Other than that, it works fine.

In fact, speaking of stable, I would point out that I added an MSRV to the README. I think it was pretty recent anyway due to the version of your `smallvec` dependency, but my `register!` macros require expression placement which was stabilized in 1.45 (i.e. earlier this month).

schungx commented:
> In fact, speaking of stable, I would point out that I added an MSRV to the README. I think it was pretty recent anyway due to the version of your `smallvec` dependency, but my `register!` macros require expression placement which was stabilized in 1.45 (i.e. the last stable release).

Yes, I was using my own `StaticVec` type, but someone else brought it up that `smallvec` already does that and is used in Servo.  So I just swapped to it instead.

schungx commented:
Question: I am trying to use the macros.  However, how do I put in generic functions?

Many standard package API's depend on a lot of different types.

schungx commented:
OK, almost immediate ran into a problem:

```rust
use rhai::codegen::{export_module, exported_module};
use rhai::{Engine, EvalAltResult, INT};

#[export_module]
pub mod array_package {
    use rhai::{Array, INT};

    pub fn len(array: &mut Array, mul: INT) -> INT {
        (array.len()) as INT * mul
    }
}

#[test]
fn test_plugins() -> Result<(), Box<EvalAltResult>> {
    let mut engine = Engine::new();

    let m = exported_module!(array_package);

    engine.load_package(m.into());

    assert_eq!(engine.eval::<INT>("let a = [1, 2, 3]; a.len(2)")?, 6);

    Ok(())
}
```

The error is `procedural macros cannot be expanded to expressions` on `exported_module!(array_package)`.

schungx commented:
OK, just realized that the test works on nightly.  Does `codegen` only work on nightly?  Or should I do something for it to work on stable?

jhwgh1968 commented:
> Question: I am trying to use the macros. However, how do I put in generic functions?
>
> Many standard package API's depend on a lot of different types.

That is one of the areas I have not tested yet. I suspect there are issues with it, which I hope are easy to fix (depending on how good generic inference like `Vec<_>` is).

I will focus on that in the coming days.

> The error is `procedural macros cannot be expanded to expressions` on `exported_module!(array_package)`.

For the record, this error is why [the MSRV is now 1.45](https://blog.rust-lang.org/2020/07/16/Rust-1.45.0.html#stabilizing-function-like-procedural-macros-in-expressions-patterns-and-statements) and why I I felt it worth documenting in the README. It struck me as a rather cryptic error to those not following Rust feature development.

I did test this on stable before pushing, and I expect it to work. (Just the `trybuild` tests fail, because they give different answers.) If you still have issues with the current stable release, please let me know.

> `export_module` seems to depend on `rhai::plugin::CallableFunction` while `export_fn` depends only on `CallableFunction`. This doesn't seem to be very consistent.

That was an oversight on my part. I will write a PR to fix it. In fact, I may be able to make both of those imports go away (by importing them myself in the generated code at the right spot).

schungx commented:
> That was an oversight on my part. I will write a PR to fix it. In fact, I may be able to make both of those imports go away (by importing them myself in the generated code at the right spot).

Great! Make sure you don't hard-code `rhai` as the root because some users may import `rhai` under an alias...

> For the record, this error is why [the MSRV is now 1.45](https://blog.rust-lang.org/2020/07/16/Rust-1.45.0.html#stabilizing-function-like-procedural-macros-in-expressions-patterns-and-statements) and why I I felt it worth documenting in the README. It struck me as a rather cryptic error to those not following Rust feature development.

Ah.  OK.  I have been running on 1.44.  Will try.

jhwgh1968 commented:
> Great! Make sure you don't hard-code `rhai` as the root because some users may import `rhai` under an alias...

Unfortunately, [I don't think that is possible to fix right now](https://www.youtube.com/watch?v=g4SYTOc8fL0#t=1247). I have already hard-coded `rhai` several places already.

schungx commented:
I wonder how other Rust libraries do it... thinks like `serde`... they must have a way to import traits and types...

jhwgh1968 commented:
I think the answer is: they require users to import everything, but keep their API small enough to make that easy. I have not tested it, but I know I often need to explicitly import `serde::{Deserialize, Serialize}` at the top of the file when using their `derive` macro for non-trivial things.

With Rhai, the API is much more complex, and I use a lot of types and traits. The only thing I could think of would be to require users import a "prelude module", the way that `rayon` and `chrono` do.

schungx commented:
> With Rhai, the API is much more complex, and I use a lot of types and traits. The only thing I could think of would be to require users import a "prelude module", the way that `rayon` and `chrono` do.

I guess forcing `use codegen::*` is OK...  Whoever wants to write modules should read the manual anyway...

jhwgh1968 commented:
I'm currently working on this, and managed to discover a nightly ICE in diagonstic messages! :laughing: 

I expect to have a PR today which at least _improves_ hygene, even if it's not perfect.

jhwgh1968 commented:
Let's continue this conversation over on #3.
